tests/integration/utils.py

```diff
@@ -463,3 +472,170 @@ def _validate_expected_dep_file_contents(dep_contents_file: Path, output_dir: Pa
         dep_file = output_dir / "deps" / path
         assert dep_file.exists()
         assert dep_file.read_text() == expected_content
+
+
+def retrieve_changed_files_from_git() -> tuple[Path, ...]:
+    repo = Repo(".", search_parent_directories=True)
+    # >>> type(repo.branches)
+    # <class 'git.util.IterableList'>
+    # Despite the fact stated above mypy does not believe one can use 'in' with
+    # repo.branches because branches is an alias to heads and heads are decorated
+    # with @property:
+    #  Unsupported right operand type for in ("Callable[[], IterableList[Head]]")
```

Comment:
side note: Digging into GitPython was a nice breeze of great Python code to have a look at :) to understand what you're saying:

> \# repo.branches because branches is an alias to heads and heads are decorated
> \# with property:

Essentially, yes, but I think it's much more poetic than that and I think you're giving mypy too much credit because I don't think mypy would be able to infer the types correctly even if `@property` weren't involved :D and it'd IMO still complain about the string. The reason for that is that `IterableList` is leveraging the `_id_attribute_` class attribute which is of type `str` and eventually translates to the `name` property because of `SymbolicReference`, but that only gets to `Head` via a triple multiple inheritance of `Reference`. For anyone interested, here are sources to infer the whole inheritance chain:

- https://github.com/gitpython-developers/GitPython/blob/cfadd9e5edb8571c2c15878c8eeca91f15a69040/git/refs/head.py#L128
- https://github.com/gitpython-developers/GitPython/blob/cfadd9e5edb8571c2c15878c8eeca91f15a69040/git/refs/reference.py#L41
- https://github.com/gitpython-developers/GitPython/blob/cfadd9e5edb8571c2c15878c8eeca91f15a69040/git/util.py#L1237
- https://github.com/gitpython-developers/GitPython/blob/cfadd9e5edb8571c2c15878c8eeca91f15a69040/git/refs/symbolic.py#L100

Reply:
True, I published a much shorter version of the story and stopped digging at an earlier point. Thank you for unraveling it further, it is an interesting story, I became mostly content with my findings when visiting reference.py. To me the most important takeaway is that mypy is incapable of doing correct inference for a legit usecase. Moreover I will never believe that it is impossible to statically infer this. I am using the word 'believe'  to denote that I am not going to prove or disprove my point with code in a language that has an actual type checker, at least not now.